src/permutations.jl

```diff
@@ -9,67 +9,121 @@ export
     permutations
 
 
-struct Permutations{T}
-    a::T
-    t::Int
+struct PermutationIterator{T}
+    data::T
+    length::Int
 end
 
-Base.eltype(::Type{Permutations{T}}) where {T} = Vector{eltype(T)}
+function has_repeats(state::Vector{Int})
+    for outer in eachindex(state)
+        for inner in (outer+1):lastindex(state)
```

Comment:
The inner loop can be marked `@inbounds` safely, I think

Reply:
The inner can be marked `@inbounds` safely because of the type restriction in the signature. It might be good to note that in a comment in case that type restriction is ever loosened.

Reply:
Thanks for the input! I've added the `inbounds`, and a comment explaining why it is safe

Reply:
Thank you. Now someone from JuliaMath needs to do a final review, merge, and release.

Reply:
Hi @mschauer, would you be willing to do the final review, merge, and release? I have reviewed it myself, and believe it is ready for the main branch.